### PR TITLE
warning letter backend first draft

### PR DESCRIPTION
**Pull request by zero-udo**
First peaces for the warning letter backend structure, described in #12.

### CODE REVIEW

File: backend/src/main/kotlin/org/darkSolace/muse/securityModule/controller/AuthController.kt [hunk 1]
```diff
@@ -19,15 +20,18 @@ import org.springframework.web.bind.annotation.RestController
...
+) {
     /**
      * Checks a transmitted [LoginRequest] for a valid username/password pair. Listens on /api/auth/signin.
      *
      * @sample `curl -X POST -H "Content-Type: application/json" -d '{ "username": "test", "password": "123" }'
      *          localhost:8000/api/auth/signin`
      * @param loginRequest a [LoginRequest] containing username and password
      * @return a [org.darkSolace.muse.securityModule.model.JwtResponse] containing
-     * a token or HTTP 401 is username or password are invalid
+     * a token, HTTP 401 is username or password are invalid, or HTTP 301 is user is suspended
```
Comment (BibiRosenberg) [inline backend/src/main/kotlin/org/darkSolace/muse/securityModule/controller/AuthController.kt:34]: HTTP 301 _if_ user is suspended

File: backend/src/main/kotlin/org/darkSolace/muse/userModule/controller/UserRestController.kt [hunk 2]
```diff
@@ -35,5 +41,69 @@ class UserRestController(@Autowired val userService: UserService) {
...
+
+    /**
+     * Deletes a user identified by its id. Listens on /api/user/{id} for DELETE requests.
+     * You need the [org.darkSolace.muse.userModule.model.Role.ADMINISTRATOR] role to access this endpoint.
+     *
+     * @sample `curl -X DELETE -H "Authorization: [...]" localhost:8080/api/user/5`
+     * @param id the user id
+     */
+    @DeleteMapping("/{id}")
+    @PreAuthorize("hasAuthority('ADMINISTRATOR')")
```
Comment (BibiRosenberg) [inline backend/src/main/kotlin/org/darkSolace/muse/userModule/controller/UserRestController.kt]: Do we need to make that available for plain users too? Most sites allow you to delete your own account, don't they?

File: backend/src/main/kotlin/org/darkSolace/muse/userModule/repository/SuspensionHistoryRepository.kt [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+import org.darkSolace.muse.userModule.model.SuspensionHistoryEntry
+import org.darkSolace.muse.userModule.model.User
+import org.springframework.data.repository.CrudRepository
+
+interface SuspensionHistoryRepository : CrudRepository<SuspensionHistoryEntry, Long> {
+    fun getById(id: Long): SuspensionHistoryEntry?
+    fun getByUserAndAcceptedDateIsNull(user: User): SuspensionHistoryEntry?
+    fun getByConfirmationCode(confirmationCode: String): SuspensionHistoryEntry?
+    fun findAllByUserOrderBySuspendedDate(user: User): List<SuspensionHistoryEntry>
```
Comment (BibiRosenberg) [inline backend/src/main/kotlin/org/darkSolace/muse/userModule/repository/SuspensionHistoryRepository.kt:10]: Please add functions to 
- find all currently suspended users
- find all with entries in SuspensionHistory (accepted and not)

File: backend/src/main/kotlin/org/darkSolace/muse/userModule/service/UserRoleService.kt [hunk 4]
```diff
@@ -49,4 +80,51 @@ class UserRoleService(@Autowired val userRepository: UserRepository) {
...
+    fun confirmSuspension(confirmationCode: String): Boolean {
+        val suspension = suspensionHistoryRepository.getByConfirmationCode(confirmationCode) ?: return false
+        changeRole(suspension.user, Role.MEMBER)
+        suspension.acceptedDate = Date()
+        suspensionHistoryRepository.save(suspension)
+        return true
+    }
+
+    /**
+     * Retrieves the suspention history of a given user, identified by its id.
```
Comment (BibiRosenberg) [inline backend/src/main/kotlin/org/darkSolace/muse/userModule/service/UserRoleService.kt]: Spell check in this file for suspention, please ;)

File: backend/src/main/kotlin/org/darkSolace/muse/userModule/model/SuspensionHistoryEntry.kt [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    @GeneratedValue
+    @Id
+    val id: Long?,
+    @OneToOne
+    val user: User,
+    @Temporal(TemporalType.TIMESTAMP)
+    val suspendedDate: Date = Date(),
+    val confirmationCode: String = UUID.randomUUID().toString(),
+    @Temporal(TemporalType.TIMESTAMP)
+    var acceptedDate: Date? = null
```
Comment (BibiRosenberg) [inline backend/src/main/kotlin/org/darkSolace/muse/userModule/model/SuspensionHistoryEntry.kt]: Could/should we add confirmationDetails, like browser metadata from the user and OS or something?

File: backend/src/main/kotlin/org/darkSolace/muse/userModule/service/UserService.kt [hunk 6]
```diff
@@ -162,14 +172,14 @@ class UserService(
      *
      * @param signUpRequest The [SignUpRequest] containing all required information
```
Comment (BibiRosenberg) [inline backend/src/main/kotlin/org/darkSolace/muse/userModule/service/UserService.kt:173]: @return User is missing

File: backend/src/test/kotlin/org/darkSolace/muse/userModule/UserControllerApiTests.kt [hunk 7]
```diff
@@ -92,19 +131,193 @@ class UserControllerApiTests {
...
+
+        url = generateUrl("/api/auth/signin")
+        val signInResponse = restTemplate.postForEntity(
+            url,
+            SignUpRequest("test2", "123", "test2@example.com"),
+            JwtResponse::class.java
+        )
+
+        //suspend user via api
+        url = generateUrl("/api/user/suspend/${user?.id}")
```
Comment (BibiRosenberg) [inline backend/src/test/kotlin/org/darkSolace/muse/userModule/UserControllerApiTests.kt:176]: did you not change this to suspend/confirmationCode/something?

**Review by BibiRosenberg — CHANGES_REQUESTED** [review]

File: backend/src/main/kotlin/org/darkSolace/muse/userModule/controller/UserRestController.kt (hunk 2 above)
Comment (zero-udo) [reply backend/src/main/kotlin/org/darkSolace/muse/userModule/controller/UserRestController.kt]: added the possibility for users to delete themselves. Other users can still be deleted by an administrator

File: backend/src/main/kotlin/org/darkSolace/muse/userModule/repository/SuspensionHistoryRepository.kt (hunk 3 above)
Comment (zero-udo) [reply backend/src/main/kotlin/org/darkSolace/muse/userModule/repository/SuspensionHistoryRepository.kt:10]: find all currently suspended users function added, find all entries is part of the CrudRepository parent class (as well as findById - which is why i removed getById as well *oops)

File: backend/src/main/kotlin/org/darkSolace/muse/userModule/model/SuspensionHistoryEntry.kt (hunk 5 above)
Comment (zero-udo) [reply backend/src/main/kotlin/org/darkSolace/muse/userModule/model/SuspensionHistoryEntry.kt]: added a string-field for confirmationDetails. The API endpoint has to changed as well when we know which data and how this data might be sent to the backend.

File: backend/src/test/kotlin/org/darkSolace/muse/userModule/UserControllerApiTests.kt (hunk 7 above)
Comment (zero-udo) [reply backend/src/test/kotlin/org/darkSolace/muse/userModule/UserControllerApiTests.kt:176]: suspending a user is done via `/api/user/suspend/<id>` - if a user confirms it suspension (after reviewing the warning letter) `/api/user/suspend/confirm/<uuid>` is called

File: backend/src/test/kotlin/org/darkSolace/muse/userModule/UserControllerApiTests.kt [hunk 8]
```diff
@@ -92,19 +131,308 @@ class UserControllerApiTests {
...
+            add("Authorization", "Bearer ${signInResponse?.body?.token}")
+        }
+        val response =
+            restTemplate.exchange(url, HttpMethod.DELETE, HttpEntity<HttpHeaders>(headers), HttpStatus::class.java)
+
+        Assertions.assertEquals(HttpStatus.OK, response.statusCode)
+
+        user = userService.getById(user?.id ?: -1)
+        Assertions.assertNotNull(user)
+    }
```
Comment (BibiRosenberg) [inline backend/src/test/kotlin/org/darkSolace/muse/userModule/UserControllerApiTests.kt:474]: please add tests for
- member trying to delete other member (or mod, or admin)
- anyone trying to delete an unknown user

**Review by BibiRosenberg — APPROVED** [review]

**Pull request merged**